datacosmos/stac/search.py

```diff
@@ -0,0 +1,257 @@
+    levels: List[Levels] | None = None
+    after: datetime.datetime | None = None
+    before: datetime.datetime | None = None
+    bbox: List[float] | None = None
+    intersects: geojson.GeoJSON | Dict | None = None
+    min_cloud_cover_pct: float | None = None
+    max_cloud_cover_pct: float | None = None
+    min_resolution_m: float | None = None
+    max_resolution_m: float | None = None
+    query: Dict | None = None
+
+    def _add_constellations(self, body: Dict) -> None:
+        if self.constellations is not None:
+            body["query"]["constellation"] = {
+                "in": [c.value for c in self.constellations]
+            }
+
+    def _add_satellites(self, body: Dict) -> None:
+        if self.satellites is not None:
+            body["query"]["sat:platform_international_designator"] = {
+                "in": [c.value for c in self.satellites]
+            }
+
+    def _add_levels(self, body: Dict) -> None:
+        if self.levels is not None:
+            body["query"]["processing:level"] = {"in": [c.value for c in self.levels]}
+
+    def _add_datetimes(self, body: Dict) -> None:
+        expr = {}
+        if self.after is not None:
+            expr["gte"] = self.after.isoformat()
+        if self.before is not None:
+            expr["lte"] = self.before.isoformat()
+        if expr:
+            body["query"]["datetime"] = expr
+
+    def _add_bbox(self, body: Dict) -> None:
+        if self.bbox is not None:
+            body["bbox"] = self.bbox
+
+    def _add_intersects(self, body: Dict) -> None:
+        if self.intersects is not None:
+            body["intersects"] = self.intersects
+
+    def _add_cloud_cover(self, body: Dict) -> None:
+        expr = {}
+        if self.min_cloud_cover_pct is not None:
+            expr["gte"] = self.min_cloud_cover_pct
+        if self.max_cloud_cover_pct is not None:
+            expr["lte"] = self.max_cloud_cover_pct
+        if expr:
+            body["query"]["eo:cloud_cover"] = expr
+
+    def _add_resolution(self, body: Dict) -> None:
+        expr = {}
+        if self.min_resolution_m is not None:
+            expr["gte"] = self.min_resolution_m
+        if self.max_resolution_m is not None:
+            expr["lte"] = self.max_resolution_m
+        if expr:
+            body["query"]["gsd"] = expr
+
+    def _add_query(self, body: Dict) -> None:
+        if self.query is not None:
+            body["query"].update(self.query)
+
+    def build(self) -> Dict:
+        """Build the search query body to be sent to the API."""
+        body = {}
+        body["query"] = {}
+
+        self._add_constellations(body)
+        self._add_satellites(body)
+        self._add_levels(body)
+        self._add_datetimes(body)
+        self._add_bbox(body)
+        self._add_intersects(body)
+        self._add_cloud_cover(body)
+        self._add_resolution(body)
+        self._add_query(body)
+
+        if not body["query"]:
+            del body["query"]
```

Comment:
Does this remove the `query` field if it's still an empty dict? I'm not sure if `not {{non-boolean type}}` is considered Pythonic, but I think the intent of the code would become clearer in many cases if we explicitly stated the condition we're checking for like

```python
if len(body["query"]) == 0:
    del body["query"]
```

First though I would ensure that sending an empty query parameter to the STAC service will actually produce bad results, if it will not, then we can omit this step altogether.

Reply:
python guidance is to avoid doing `if len` on containers, all empty containers count as false. https://peps.python.org/pep-0008/#programming-recommendations

I agree we can get rid of this though

Reply:
done